discord/commands/mixins.py

```diff
@@ -0,0 +1,952 @@
+        except AttributeError:
+            pass
+        return other
+
+    def _update_copy(self, kwargs: dict[str, Any]):
+        if kwargs:
+            kw = kwargs.copy()
+            kw.update(self.__original_kwargs__)
+            copy = self.__class__(self.callback, **kw)
+            return self._ensure_assignment_on_copy(copy)
+        else:
+            return self.copy()
+
+    def _set_cog(self, cog: CogT):
+        self.cog = cog
+
+    def is_on_cooldown(self, ctx: BaseContext) -> bool:
+        """Checks whether the command is currently on cooldown.
+
+        .. note::
+
+            This uses the current time instead of the interaction time.
+
+        Parameters
+        ----------
+        ctx: :class:`.BaseContext`
+            The invocation context to use when checking the command's cooldown status.
+
+        Returns
+        -------
+        :class:`bool`
+            A boolean indicating if the command is on cooldown.
+        """
+        if not self._buckets.valid:
+            return False
+
+        bucket = self._buckets.get_bucket(ctx)
+        current = utils.utcnow().timestamp()
+        return bucket.get_tokens(current) == 0
+
+    def reset_cooldown(self, ctx) -> None:
+        """Resets the cooldown on this command.
+
+        Parameters
+        ----------
+        ctx: :class:`.BaseContext`
+            The invocation context to reset the cooldown under.
+        """
+        if self._buckets.valid:
+            bucket = self._buckets.get_bucket(ctx)  # type: ignore # ctx instead of non-existent message
+            bucket.reset()
+
+    def get_cooldown_retry_after(self, ctx) -> float:
+        """Retrieves the amount of seconds before this command can be tried again.
+
+        .. note::
+
+            This uses the current time instead of the interaction time.
+
+        Parameters
+        ----------
+        ctx: :class:`.BaseContext`
+            The invocation context to retrieve the cooldown from.
+
+        Returns
+        -------
+        :class:`float`
+            The amount of time left on this command's cooldown in seconds.
+            If this is ``0.0`` then the command isn't on cooldown.
+        """
+        if self._buckets.valid:
+            bucket = self._buckets.get_bucket(ctx)
+            current = utils.utcnow().timestamp()
+            return bucket.get_retry_after(current)
+
+        return 0.0
+
+    def _prepare_cooldowns(self, ctx: BaseContext):
+        if not self._buckets.valid:
+            return
+
+        current = datetime.datetime.now().timestamp()
+        bucket = self._buckets.get_bucket(ctx, current)  # type: ignore # ctx instead of non-existent message
+
+        if bucket:
+            retry_after = bucket.update_rate_limit(current)
+
+            if retry_after:
+                raise CommandOnCooldown(bucket, retry_after, self._buckets.type)  # type: ignore
+
+    async def call_before_hooks(self, ctx: BaseContext) -> None:
```

Comment:
Docstring? If it's not supposed to be public, then make it private via a trailing underscore.